src/algebra/group_with_zero.lean

```diff
@@ -0,0 +1,504 @@
+by simp [h]
+
+lemma one_div_mul_cancel {a : G} (h : a ≠ 0) : (1 / a) * a = 1 :=
+by simp [h]
+
+lemma one_div_one : 1 / 1 = (1:G) :=
+div_self (ne.symm zero_ne_one)
+
+lemma one_div_ne_zero {a : G} (h : a ≠ 0) : 1 / a ≠ 0 :=
+assume : 1 / a = 0,
+have 0 = (1:G), from eq.symm (by rw [← mul_one_div_cancel h, this, mul_zero]),
+absurd this zero_ne_one
+
+lemma mul_ne_zero_comm {a b : G} (h : a * b ≠ 0) : b * a ≠ 0 :=
+by { rw mul_ne_zero_iff at h ⊢, rwa and_comm }
+
+lemma eq_one_div_of_mul_eq_one {a b : G} (h : a * b = 1) : b = 1 / a :=
+have a ≠ 0, from
+   assume : a = 0,
+   have 0 = (1:G), by rwa [this, zero_mul] at h,
+      absurd this zero_ne_one,
+have b = (1 / a) * a * b, by rw [one_div_mul_cancel this, one_mul],
+show b = 1 / a, by rwa [mul_assoc, h, mul_one] at this
+
+lemma eq_one_div_of_mul_eq_one_left {a b : G} (h : b * a = 1) : b = 1 / a :=
+have a ≠ 0, from
+  assume : a = 0,
+  have 0 = (1:G), by rwa [this, mul_zero] at h,
+    absurd this zero_ne_one,
+by rw [← h, mul_div_assoc, div_self this, mul_one]
+
+@[simp] lemma one_div_div (a b : G) : 1 / (a / b) = b / a :=
+by rw [one_div, div_eq_mul_inv, mul_inv_rev, inv_inv, div_eq_mul_inv]
+
+@[simp] lemma one_div_one_div (a : G) : 1 / (1 / a) = a :=
+by simp
+
+lemma eq_of_one_div_eq_one_div {a b : G} (h : 1 / a = 1 / b) : a = b :=
+by rw [← one_div_one_div a, h, one_div_one_div]
+
+end gwz
+
+section not_comm
+
+variables {G : Type*} [group_with_zero G]
+variables {a b c : G}
+
+@[simp] lemma inv_eq_zero {a : G} : a⁻¹ = 0 ↔ a = 0 :=
+classical.by_cases (assume : a = 0, by simp [*]) (assume : a ≠ 0, by simp [*, gwz.inv_ne_zero])
+
+lemma one_div_mul_one_div_rev (a b : G) : (1 / a) * (1 / b) =  1 / (b * a) :=
+by simp only [div_eq_mul_inv, one_mul, gwz.mul_inv_rev]
+
+theorem divp_eq_div (a : G) (u : units G) : a /ₚ u = a / u :=
+congr_arg _ $ units.inv_eq_inv _
+
+@[simp] theorem divp_mk0 (a : G) {b : G} (hb : b ≠ 0) :
+  a /ₚ units.mk0 b hb = a / b :=
+divp_eq_div _ _
+
+lemma inv_div : (a / b)⁻¹ = b / a :=
+(gwz.mul_inv_rev _ _).trans (by rw gwz.inv_inv; refl)
+
+lemma inv_div_left : a⁻¹ / b = (b * a)⁻¹ :=
+(gwz.mul_inv_rev _ _).symm
+
+lemma div_ne_zero (ha : a ≠ 0) (hb : b ≠ 0) : a / b ≠ 0 :=
+gwz.mul_ne_zero ha (gwz.inv_ne_zero hb)
+
+lemma div_ne_zero_iff (hb : b ≠ 0) : a / b ≠ 0 ↔ a ≠ 0 :=
+⟨mt (λ h, by rw [h, gwz.zero_div]), λ ha, div_ne_zero ha hb⟩
+
+lemma div_eq_zero_iff (hb : b ≠ 0) : a / b = 0 ↔ a = 0 :=
+by haveI := classical.prop_decidable; exact
+not_iff_not.1 (div_ne_zero_iff hb)
+
+lemma div_right_inj (hc : c ≠ 0) : a / c = b / c ↔ a = b :=
+by rw [← divp_mk0 _ hc, ← divp_mk0 _ hc, divp_right_inj]
+
+lemma gwz.mul_right_inj (hc : c ≠ 0) : a * c = b * c ↔ a = b :=
+by rw [← gwz.inv_inv c, ← div_eq_mul_inv, ← div_eq_mul_inv, div_right_inj (gwz.inv_ne_zero hc)]
+
+lemma div_eq_iff_mul_eq (hb : b ≠ 0) : a / b = c ↔ c * b = a :=
+⟨λ h, by rw [← h, gwz.div_mul_cancel _ hb],
+ λ h, by rw [← h, gwz.mul_div_cancel _ hb]⟩
+
+end not_comm
+
+namespace gwz -- comm
```

Comment:
I would put these in the `comm_group_with_zero` namespace. Even with dot notation irrelevant, I think it's important that lemmas live in the right namespace.

Reply:
I'm pretty glad that most algebraic lemmas live in the root namespace, because it would be a nightmare guessing which algebraic structure each of the lemmas live in. If I'm writing a theorem about ordered rings I might use `add_zero` which is about add_monoids and `mul_lt_mul` which is about ordered semirings, and `mul_le_mul_left` which is about linear ordered semirings.

So I sympathize with the idea to keep the namespaces down here. I think namespaces should only be used here for disambiguation, when there is another theorem of the same name (and roughly the same statement) on an incompatible typeclass, and the namespace should just provide that disambiguating information. That said, `gwz` is not descriptive, and there are no easy ways to read a doc string about what the namespace is abbreviating. 

Another alternative, which is used in a few places like the theorems about `ordered_comm_group` in `algebra.ordered_group`, is to prime the theorem name. This is not descriptive, but it avoids the namespace overhead, and these are intended to be browsed via an autocomplete menu anyway, and the statements should disambiguate them.

Reply:
I can switch to `foobar'` everywhere.